cmd/pinniped-proxy/src/pinniped.rs

```diff
@@ -78,26 +77,199 @@ fn identity_for_exchange(cred: &ClusterCredential) -> Result<Identity> {
 ///
 /// The rust derive macro together with the kube macro creates serializable and deserializable
 /// resources based on the struct. See https://docs.rs/kube/0.43.0/kube/ for more details.
-#[derive(CustomResource, Deserialize, Serialize, Clone, Debug)]
-#[kube(group = "login.concierge.pinniped.dev", version = "v1alpha1", kind = "TokenCredentialRequest", namespaced)]
-#[kube(status = "TokenCredentialRequestStatus")]
+
+
+// From https://github.com/clux/kube-rs/blob/master/kube-derive/src/lib.rs
```

Comment:
> here's only two types there: `str` and `String`. The `'static` is just a lifetime specifier telling the compiler that that variable should live for as long as the program itself (ie. don't free it when it goes out of scope, which is what the compiler will do by default for a stack-allocated var unless ownership is passed on to another piece of code).
> 
> Yep, thinking about ownership is the difficult part. In this case, an `str` is always passed by reference (ie. an `&str`, since it's just a pointer to a slice of memory somewhere, not a struct). Where as a `String` is an struct that handles pushing extra characters, joining and a bunch of other things, that can be owned or borrowed by others as you decide.

I'm getting the point. After having a look at it again, the problem was not a type problem, but a misunderstanding between `static` and `const`. The latter is constant but at compile time! So I was trying to join string, but it is not supported.
Indeed, I think here is where the original problem (use the derive macro) strikes back...

```rust
// It is required to implement this :(
impl k8s_openapi::Resource for TokenCredentialRequest {
    // Not compiling
    const API_VERSION: &'static str = format!(&get_pinniped_login_api_group(), &DEFAULT_VERSION).as_str();
    // Not compiling
    const GROUP: &'static str = &get_pinniped_login_api_group();
    const KIND: &'static str = "TokenCredentialRequest";
    const VERSION: &'static str = &DEFAULT_VERSION;
}
```
`calls in constants are limited to constant functions, tuple structs and tuple variants.`

I have taken a look at https://crates.io/crates/lazy_static, but still not working...
Another idea is to use the cargo build scripts, but the problem is yet the same... not able to change these things in runtime.
I'm running out of new approaches to deal with that; I'm starting to think the best option is just to support a fixed number of API suffixes as in #2536 :S

Reply:
> I'm running out of new approaches to deal with that; I'm starting to think the best option is just to support a fixed number of API suffixes as in #2536 :S

Yes, I thinkn the two API suffixes is fine for now. I'd like to have a look at generalising it when I have time, but it's not urgent atm.